계산기 II [STEP2] borysarang, yeton

yusw10 opened the pull request:
안녕하세요 코리! @corykim0829   계산기2 step2 PR 보냅니다.

이번 프로젝트는 대부분 오류 처리 하는 데 시간을 썼습니다.
아래는 변경되거나 추가된 파일(타입)을 적어주었습니다.


[변경해준 파일(타입)]

1. ViewController
    - operandButtonDidTap 기능 추가 및 수정
        - 숫자 입력 시 20자 제한
        - 입력할 때부터 decimal이 적용되도록 하기
        - 불필요한 조건문을 String+Extension의 switch-case문을 통해 처리
    - operatorButtonDidTap 기능 수정
        - NaN일 때 연산자 입력 안되도록 수정
    - equalButtonDidTap 기능 수정
        - =를 눌렀을 때 바로 다음에 피연산자가 입력되지 않도록 수정
    - applyNumberFormatter 기능 수정
        - numberFormatter.maximumFractionDigits = 20 추가
    - applyNumberFormatterToInteger 메서드 추가
        - "1,231.000" 일 경우 정수부분인 1231은 numberFormatter가 적용되어 1,231이  하고, 소수부분은 적용되지 않도록 해주는 메서드 추가
    - isZeroNumber 메서드 추가
        - 현재 라벨의 text가 0일 때 조건들을 체크해주는 메서드 추가
        - 
2. String+Extension
    - canInput(currentOperandLabel: String?) 메서드 추가
        - 현재 operand와 입력받은 버튼의 숫자에 따라 변경하지 않는 경우의 수를 확인하고 Bool타입을 return 해줌
    - canModify(currentOperandLabel: String?) -> String?
        - 현재 operand와 입력받은 버튼의 숫자에 따라 변경하지 않는 경우의 수를 확인하고 변경할 값을 return 해줌
    - scrollToBottom 메서드 수정
        - layoutIfNeeded()를 추가하여 라벨이 동기 처리 되도록 수정

3. StackView+Extension
    - canAddOperand(_ label: String) 메서드 추가
        - =을 눌렀을 때의 경우인지를 판단하여 Bool 타입으로 return하여 숫자 라벨에 더이상 숫자가 입력되지 못하도록 해주는 메서드 추가
        - 
4. NameSpace
    - nan,comma 케이스 추가


[Step2에서 처리해준 오류]
- 숫자 입력 시 20자 제한 
- 입력할 때부터 decimal이 적용되도록 하기 ( 1,000,000,000)
- NaN일 때 연산자 입력 못받도록 하기 
- 스크롤 뷰에 올라가는 Label이 한 줄씩 밀리는 상황 발생 -> 동기 처리해줌



[배운 개념]
- string+Extension과 stackView+Extension 내부에서 조건을 처리해줄 수 있는 switch-case문을 만들어서 바로 ViewController에서 이 메서드들을 호출해서 처리해줄 수 있다는 것을 알게 됨
- switch-case에서 튜플 형태로 case를 비교할 수 있다는 것을 배움
- switch-case에서 case로 (1...9) 형식을 사용할 수 있다는 것을 배움
- switch-case에서 case let을 사용하여 조건을 걸어줄 수 있다는 것을 배움


이상 여기까지였습니다!
계산기 프로젝트 좋은 방향으로 이끌어주셔서
정말 감사합니다.

마지막 PR도 잘부탁드려요!!

yusw10 committed:
refactor: 20자 제한 기능 구현

- 메서드(changeMecimalStyle)의 numberFormatter 20자 제한 속성 부여
- 메서드(pressOperandButton) 최대 입력 제한 조건 추가

yusw10 committed:
feat: 프로퍼티(currentOperandLabel)을 decimal style로 변경

- 메서드(pressOperandButton) 20자 제한 조건문의 임시 프로퍼티 명 변경
- 메서드(pressOperandButton)에서 프로퍼티(currentOperandLabel)의 텍스트로 조건 확인시 메서드(changeDoubleStyle)의 결과값으로 비교하게 변경
- 메서드(pressOperatorButton)에서 프로퍼티(currentOperandLabel)의 텍스트로 스크롤뷰 컨텐츠 생성 확인시 메서드(changeDoubleStyle)의 결과값으로 비교하게 변경
- 메서드(pressEqualButton)에서 프로퍼티(currentOperandLabel)의 텍스트값을 연산자 스택 프로퍼티(operaionStack)에 삽입 시 메서드(changeDoubleStyle)의 결과값으로 비교하게 변경

yeeton37 committed:
feat: String을 extension하여 메서드(canInput)과 메서드(canReturn) 선언

- 계산기 스크롤뷰에 라벨이 바로 적용되도록 layoutIfNeeded 사용해줌
- 메서드(canInput), 메서드(canReturn): operand에 대한 에러 처리 함수

yusw10 committed:
refactor: String Extention의 메서드(canInput) 수정

- 스위치에 들어가는 전달인자(currentoperand)와 self의 튜플 조건식 수정

yusw10 committed:
refactor: 메서드(pressOperandButton) 기능 분리

- 0과 . 으로만 이루어진지 판별하기 위한 메서드(isZeroNumber) 추가
- 정수자리만 NumberFormatter를 적용시키기 위해 메서드(applyNumberFormatterToInteger) 추가
- 메서드(pressOperandButton) 불필요한 조건식 수정
- 내부 프로퍼티(numberFormmater) 변수명 수정

yeeton37 committed:
feat: UIStackView,String extension 및 메서드(pressOperandButton) 수정

- String+Extension: 케이스 (CalcAccessory.dot, operand), ("1"..."9", CalcAccessory.nan) 추가
- UIStackView+Extension: =을 누르고 숫자 입력 못하도록 하는 메서드(canAdd) 선언

yusw10 committed:
refactor: 변수 및 메서드 네이밍 수정

corykim0829 commented on Calculator/Calculator/BasicTypeExtension/String+Extension.swift:
```diff
@@ -0,0 +1,49 @@
+    }
+}
+
+extension String { //self , currentoperand
```
String에 extension 함수를 사용하는 시도는 아주 좋습니다!
하지만 String 모든 타입에서 이 함수를 접근할 수 있으니 한번 사용을 위해서 extension을 사용하는 것은 조금 아쉬운 것 같습니다.

corykim0829 commented on Calculator/Calculator/ViewController.swift:
```diff
@@ -179,15 +176,46 @@ extension ViewController {
+    }
+    
+    private func changeDoubleStyle(_ input: String?) -> String {
+        return input?.replacingOccurrences(of: CalcAccessory.comma, with: CalcAccessory.empty) ?? CalcAccessory.empty
```
`??`가 이렇게 한줄이 길어지면 가독성을 떨어트릴 수 있습니다. 이 전에 optional 처리를 해서 처리하면 좋을 것 같아요!

corykim0829 commented on Calculator/Calculator/ViewController.swift:
```diff
@@ -179,15 +176,46 @@ extension ViewController {
-        let numberFormmater = NumberFormatter()
-        numberFormmater.numberStyle = .decimal
-        numberFormmater.maximumSignificantDigits = 20
+    private func applyNumberFormatter(_ input: String?) -> String {
```
이 부분을 String extension으로 처리하면 더 깔끔할 것 같아요!

corykim0829 commented on Calculator/Calculator/ViewController.swift:
```diff
@@ -179,15 +176,46 @@ extension ViewController {
+        }
+    }
+    
+    private func changeDoubleStyle(_ input: String?) -> String {
```
함수명을 보면 반환값이 없는 함수 같고 어떤 기능을 하는지 유추하기가 어려워요. 여기도 String extension으로 처리하면 더 깔끔할 것 같아요.
여기는 사실상 String의 컴마를 없애기 때문에 해당 기능을 표현하는 함수명이면 좋을 것 같아요!

yusw10 replied:
말씀주신대로 removeComma 메서드로 변경하면서 수정 해보았습니다..!

[8265f33](https://github.com/yagom-academy/ios-calculator-app/pull/304/commits/8265f334b857c06ac1294786697e89f4c1f289fc)

corykim0829 commented on Calculator/Calculator/ViewController.swift:
```diff
@@ -21,48 +20,46 @@ class ViewController: UIViewController {
-            default:
-                currentOperandLabel.text = CalcAccessory.empty
-            }
+        let entireOperandLabel = changeDoubleStyle(currentOperandLabel.text) + (sender.currentTitle ?? CalcAccessory.empty)
```
변수명이 Label로 끝나면, UILabel 타입일 것 같아요. string 타입 같은 변수명으로 바꾸면 좋을 것 같습니다.

corykim0829 commented on Calculator/Calculator/BasicTypeExtension/String+Extension.swift:
```diff
@@ -0,0 +1,49 @@
+        }
+    }
+    
+    func canModify(currentOperandLabel: String?) -> String? {
```
함수명 canModify만 보면, Bool 타입을 반환할 것 같은데, String을 반환하네요!

corykim0829 replied:
또 canModify가 사용되는 곳을 보면
string.canModify(currentOperandLabel: currentOperandLabel)
문자열 값이 파라미터 값을 수정할 수 있는지 여부를 파악하는 것 같아서 어색합니다.

yusw10 replied:
수정 완료했습니다..!

[f8a05f9](https://github.com/yagom-academy/ios-calculator-app/pull/304/commits/f8a05f935bb13e03cd3e34b50c0c4de505a42eea)

corykim0829 commented on Calculator/Calculator/ViewController.swift:
```diff
@@ -21,48 +20,46 @@ class ViewController: UIViewController {
+        let rawOperandLabel = changeDoubleStyle(currentOperandLabel.text)
+        if (sender.currentTitle ?? CalcAccessory.empty).canInput(currentOperandLabel: rawOperandLabel) == false {
+            return
+        } else if let number = sender.currentTitle?.canModify(currentOperandLabel: rawOperandLabel) {
```
canModify도 string의 extension에서 사용하지 않아도 될 것 같습니다!

yusw10 replied:
위와 같이 수정 해보았습니다.

[f8a05f9](https://github.com/yagom-academy/ios-calculator-app/pull/304/commits/f8a05f935bb13e03cd3e34b50c0c4de505a42eea)

corykim0829 reviewed (changes requested):
수고하셨습니다!

extension을 적극적으로 활용하신건 좋은 시도였습니다! 하지만 결과적으로 가독성이 많이 떨어지게 되어서 아쉬웠습니다.
수정하면서 더 개선해보면 좋을 것 같습니다!!

yusw10 committed:
refactor: 메서드(canAddOperand) 리팩터링

- 메서드(canAddOperand)를 UIStackview의 익스텐션에서 뷰 컨트롤러의 내부 메서드로 변경
- 전달인자를 StackView를 받도록 수정

yeeton37 committed:
refactor: String extension 내부 메서드 수정

- String extension 내부 메서드(canInput)의 파라미터 수정
- String extension 내부 메서드(canModify) 삭제 후 ViewController 내부에서 새로운 메서드(processSpecificCase) 생성

yusw10 committed:
style: 작성자명 수정

- "Yeton, Borysarang"으로 코드 작성자 수정

yusw10 committed:
refactor: 메서드(applyNumberFormatter) 리팩터링

- 뷰컨트롤러의 내부 메서드에서 String의 Extension으로 변경
- 전달인자를 제거하고 self로 접근하도록 뷰 컨트롤러 코드 수정

yusw10 committed:
refactor: 옵셔널 바인딩 및 메서드(removeComma) 리팩터링

- 메서드(removeComma)의 선언을 뷰 컨트롤러의 내부 메서드에서 String의 Extension으로 수정
- 메서드(removeComma)를 호출하는 부분을 String Extension에서 선언한것에 맞게 변경
- 메서드(operandButtonDidTap)의 내부 프로퍼티 옵셔널 바인딩 코드 정리
- 메서드(operatorButtonDidTap)의 내부 프로퍼티 옵셔널 바인딩 코드 정리
- 메서드(processSpecificCase)의 전달인자 프로퍼티 타입 변경

yusw10 commented on Calculator/Calculator/ViewController.swift:
```diff
@@ -21,48 +20,46 @@ class ViewController: UIViewController {
-        
-        if sender.currentTitle == CalcAccessory.dot, currentOperandLabel.text?.contains(CalcAccessory.dot) == true {
+    @IBAction func operandButtonDidTap(_ sender: UIButton) {
+        if operationStackView.canAddOperand(currentOperatorLabel.text ?? CalcAccessory.empty) == false {
```
말씀주신대로 변경해 보았습니다..!

[7d2328d](https://github.com/yagom-academy/ios-calculator-app/pull/304/commits/7d2328d2177027a450f5fd5b6ffa78c31ed79efd)

yusw10 commented on Calculator/Calculator/ViewController.swift:
```diff
@@ -21,48 +20,46 @@ class ViewController: UIViewController {
-        if operandLabel == CalcAccessory.zero,
-            (sender.currentTitle == CalcAccessory.zero || sender.currentTitle == CalcAccessory.doubleZero) {
+        let rawOperandLabel = changeDoubleStyle(currentOperandLabel.text)
+        if (sender.currentTitle ?? CalcAccessory.empty).canInput(currentOperandLabel: rawOperandLabel) == false {
```
말씀 주신대로 현재라벨과 버튼의 타이틀을 변경해보았습니다!

[f8a05f9](https://github.com/yagom-academy/ios-calculator-app/pull/304/commits/f8a05f935bb13e03cd3e34b50c0c4de505a42eea)
